zproject/computed_settings.py

```diff
@@ -375,7 +378,7 @@ def get_dirs(self) -> List[str]:
 # which has its own buckets separate from the default backend.
 # In principle, it should be impossible to make requests to tornado that fall into
 # other domains, but we use this list as an extra precaution.
```

Comment:
I confess I don't fully understand why Tornado gets its own buckets, but this seems like its own decision orthogonal to the changes here.

Reply:
It's an optimization, the redis rate limiting mechanism was showing up as taking too much time in profiling of Tornado requests - and because a specific User's Tornado requests are always handled by the same process, it was possible to  implement an in-memory system. For background: https://github.com/zulip/zulip/issues/13913 https://github.com/zulip/zulip/pull/14091

Reply:
Ah -- thanks for that pointer!  That seems like it'd be useful to note in here, or as a comment on `TornadoInMemoryRateLimiterBackend`.

As I think an earlier commit in this series notes, we don't currently pas in any domain value _other_ than `api_by_user` currently; and we're repeating the pattern by only ever using `api_by_ip` with the `RateLimitedIPAddr` class.  I guess I'm not convinced that the "extra precaution" by having this variable here is doing much other than muddying the issue.

Reply:
There was some discussion on this on the tornado rate limiting PR - basically, there was concern regarding the logic choosing the rate limiting backend on `settings.RUNNING_INSIDE_TORNADO` value - and checking the `domain` was done as hardening to prevent this turning into a bug in the course of future changes:
```
class RateLimitedUser(RateLimitedObject):
    def __init__(self, user: UserProfile, domain: str='api_by_user') -> None:
        self.user = user
        self.domain = domain
        if settings.RUNNING_INSIDE_TORNADO and domain in settings.RATE_LIMITING_DOMAINS_FOR_TORNADO:
            backend: Optional[Type[RateLimiterBackend]] = TornadoInMemoryRateLimiterBackend
        else:
            backend = None
```
Original comment thread: https://github.com/zulip/zulip/pull/14091#discussion_r408452241
Plus Tim's comment suggesting a similar design to the one that ended up being used - https://github.com/zulip/zulip/pull/14091#issuecomment-614202868
Commit https://github.com/zulip/zulip/commit/13c3eaf086848ed62e62e205d98b946c4428cda1

Reply:
Yup, I read that conversation.  My statement is that I think the added variables are making the code less legible than the value that they add.

Specifically if the goal is to:
> Adding validation to ensure that the Tornado in-process rate-limiter is only used for the counters where we think it's appropriate to.

The control of "which counters we think it's appropriate to" is defined via "the two classes that check RUNNING_INSIDE_TORNADO and set backend accordingly" since _we are not using the domain concept in any way_ for those.  The "domain" is just adding additional confusion here, not any value, IMO.  Particularly because one might, at first reading, think that it refers to _domain names_!

Reply:
@alexmv Hmm so to clarify  - would your proposal be to just get rid of the "domains" stuff, also as an argument to RateLimitedUser?

Reply:
Yup, pretty much.